stdlib/public/core/UnsafeRawPointer.swift.gyb

```diff
@@ -0,0 +1,557 @@
+  public func copyBytes(from source: UnsafeRawPointer, count: Int) {
+    _debugPrecondition(
+      count >= 0, "UnsafeMutableRawPointer.copyBytes with negative count")
+    if UnsafeRawPointer(self) < source
+       || (source + count) <= UnsafeRawPointer(self) {
+      // copy forward from a disjoint or following overlapping range.
+      Builtin.int_memcpy_RawPointer_RawPointer_Int64(
+        _rawValue, source._rawValue, UInt64(count)._value,
+        /*alignment:*/ Int32(0)._value,
+        /*volatile:*/ false._value)
+    }
+    else {
+      // copy backward from a non-following overlapping range.
+      var destPtr = self + (count - 1)
+      var srcPtr = source + (count - 1)
+      while destPtr >= self {
+        Builtin.int_memcpy_RawPointer_RawPointer_Int64(
+          destPtr._rawValue, srcPtr._rawValue, Int64(1)._value,
+          /*alignment:*/ Int32(0)._value,
+          /*volatile:*/ false._value)
+        destPtr -= 1
+        srcPtr -= 1
+      }
+    }
+  }
+%  end # mutable
+
+  //
+  // Protocol conformance
+  //
+
+  /// The hash value.
+  ///
+  /// **Axiom:** `x == y` implies `x.hashValue == y.hashValue`.
+  ///
+  /// - Note: The hash value is not guaranteed to be stable across
+  ///   different invocations of the same program.  Do not persist the
+  ///   hash value across program runs.
+  public var hashValue: Int {
+    return Int(Builtin.ptrtoint_Word(_rawValue))
+  }
+
+  /// Return `end - self`.
+  public func distance(to x: ${Self}) -> Int {
+    return x - self
+  }
+
+  /// Return `self + n`.
+  public func advanced(by n: Int) -> ${Self} {
+    return self + n
+  }
+}
+
+@_transparent
+public func == (lhs: ${Self}, rhs: ${Self}) -> Bool {
+  return Bool(Builtin.cmp_eq_RawPointer(lhs._rawValue, rhs._rawValue))
+}
+
+@_transparent
+public func < (lhs: ${Self}, rhs: ${Self}) -> Bool {
+  return Bool(Builtin.cmp_ult_RawPointer(lhs._rawValue, rhs._rawValue))
+}
+
+@_transparent
+public func + (lhs: ${Self}, rhs: Int) -> ${Self} {
+  return ${Self}(Builtin.gep_Word(
+      lhs._rawValue, rhs._builtinWordValue))
+}
+
+@_transparent
+public func + (lhs: Int, rhs: ${Self}) -> ${Self} {
+  return rhs + lhs
+}
+
+@_transparent
+public func - (lhs: ${Self}, rhs: Int) -> ${Self} {
+  return lhs + -rhs
+}
+
+@_transparent
+public func - (lhs: ${Self}, rhs: ${Self}) -> Int {
+  return
+    Int(Builtin.sub_Word(Builtin.ptrtoint_Word(lhs._rawValue),
+                         Builtin.ptrtoint_Word(rhs._rawValue)))
+}
+
+@_transparent
+public func += (lhs: inout ${Self}, rhs: Int) {
+  lhs = lhs + rhs
+}
+
+@_transparent
+public func -= (lhs: inout ${Self}, rhs: Int) {
+  lhs = lhs - rhs
+}
+
+extension Unsafe${Mutable}RawPointer : CustomDebugStringConvertible {
+  /// A textual representation of `self`, suitable for debugging.
+  public var debugDescription: String {
+    return _rawPointerToString(_rawValue)
+  }
+}
+
+extension Unsafe${Mutable}RawPointer : CustomReflectable {
+  public var customMirror: Mirror {
```

Comment:
Tests that are the same for UnsafeRawPointer and UnsafePointer are already inside UnsafePointer.switft.gyb. This file is only for tests that are unique to raw pointer.